### PR TITLE
Support for SDPA for SAM models

### PR DESCRIPTION
**Pull request by MagnusS0**
# What does this PR do?

This PR adds support for SDPA to the SAM (Segment Anything Model), following the implementation done by the PyTorch team in [segment-anything-fast](https://github.com/pytorch-labs/segment-anything-fast/).

I add two new classes:

- `SamSdpaAttention(SamAttention)`: For the decoder.
- `SamSdpaVisionAttention(SamVisionAttention)`: For the encoder.

### Performance Improvements (Updated 03.11.2024)
Time Reduction between Eager and SDPA modes for SAM-B and SlimSAM averaged across 10 runs in ms:

| **Batch Size** | **Eager (ms)**      | **SDPA (ms)**       | **Time Reduction (%)** |
|----------------|---------------------|---------------------|------------------------|
| **SAM-B**      |                     |                     |                        |
| 1              | 64.58 (±2.14)       | 30.95 (±0.07)       | 52.06%                 |
| 8              | 1,021.45 (±4.55)    | 227.44 (±0.67)      | 77.73%                 |
| **SlimSAM**    |                     |                     |                        |
| 1              | 50.51 (±1.19)       | 28.70 (±0.03)       | 43.16%                 |
| 8              | 913.97 (±6.73)      | 284.43 (±2.14)      | 68.87%                 |

Peak Memory Usage During Forward Pass:

| **Batch Size** | **Eager (MB)** | **SDPA (MB)** | **Memory Reduction (%)** |
|----------------|----------------|---------------|--------------------------|
| **SAM-B**      |                |               |                          |
| 1              | 2,151          | 875           | 59.31%                   |
| 8              | 16,007         | 3,913         | 75.54%                   |
| **SlimSAM**    |                |               |                          |
| 1              | 1,961          | 1,273         | 35.10%                   |
| 8              | 15,540         | 9,503         | 38.84%                   |

On the TrashNet 1.0 validation set I got a reduction around 30-35%, with no decrease in accuracy.
I also ran training on TrashNet 1.0 with no issues and about 1.7x speedup per epoch with SlimSAM. Also allowed gradient check-pointing that seemed to be missing from the model deceleration, worked without any changes.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@amyeroberts 
@qubvel 
I'll also tag @ArthurZucker since we just worked on this model last weekend.

### Some TODO
Should probably add some documentation for this, but let me know where is best place to add this.

### CODE REVIEW

File: src/transformers/models/sam/modeling_sam.py
Comment (ArthurZucker) [inline src/transformers/models/sam/modeling_sam.py:912]: this should be the same as the parrent one no?

**Review by ArthurZucker — COMMENTED** [review]
Looks good! 🤗

File: src/transformers/models/sam/modeling_sam.py
Comment (ArthurZucker) [reply src/transformers/models/sam/modeling_sam.py:912]: Oh nice then let's add a small comment about that!

**Review by ArthurZucker — APPROVED** [review]
Thanks for adding, I think this should be automatically tests, but make sure to run the slow tests

File: src/transformers/models/sam/modeling_sam.py
Comment (MagnusS0) [reply src/transformers/models/sam/modeling_sam.py:912]: Added a small comment about it in d486e627c64c6982ada52805de75857c86ad8c74

**Comment by ArthurZucker** [comment]
```FAILED tests/models/sam/test_modeling_sam.py::SamModelTest::test_sdpa_can_dispatch_non_composite_models - ValueError: The SDPA model should have SDPA attention layers``` can't merge until it's fixed!

**Comment by MagnusS0** [comment]
Apologies for the commit history mess. I had to rebase to clean things up, so the branch should now be up to date with the main branch and tests pass.

**Comment by qubvel** [comment]
Sorry, it should be the last commit, so I can approve the run 🙂

**Comment by MagnusS0** [comment]
> Sorry, it should be the last commit, in that case I can approve the run 🙂

Ahh no problem, will commit again.  Just fyi there are 4 slow tests I expect will fail, as they also do that on main.

**Comment by qubvel** [comment]
It looks like everything is fine in CI 🤗 Let's wait for a comment from @zucchini-nlp regarding setting attention implementation in config and I suppose it's ready to be merged

**Comment by MagnusS0** [comment]
Awesome 🤗  Probably just some small differences in my local setup causing the test to fail then.

File: src/transformers/models/sam/configuration_sam.py
Comment (zucchini-nlp) [reply src/transformers/models/sam/configuration_sam.py:304]: @MagnusS0 hey, no need to do this now, as each sub-config should get its attn implementation set when loading `from_pretrained()`. The to check correctness are `test_attn_implementation_composite_models` and `test_sdpa_can_dispatch_composite_models` and `test_flash_attn_2_can_dispatch_composite_models`  

Also, can you pls add the attributes for `sub_configs` as in this PR (https://github.com/huggingface/transformers/pull/34410/files) and enable tests for configs?

File: tests/models/sam/test_modeling_sam.py
Comment (zucchini-nlp) [inline tests/models/sam/test_modeling_sam.py:460]: it should be the `composite` model as SAM has several sub configs in the main config. You can add class attribute `_is_composite=True` to the tester class so the appropriate tests are called :)

Comment (zucchini-nlp) [reply tests/models/sam/test_modeling_sam.py]: cool, thanks, LGTM! Feel free to ping core maintainer for review :)

**Review by ArthurZucker — COMMENTED** [review]
Thanks a lot for the addition, sorry for the delay. 
We are going to refactor a bit afterwards in #35235, but let's merge for now! 🤗

**Pull request merged**
